Add media query hooks and animation to snackbar notices

youknowriad opened the pull request:
This PR adds some animation to snackbar notices. It uses `react-spring` to do so and it introduces new React hooks: `useMediaQuery` and `useReducedMotion` in order to disable the animation if the reduced motion preference is enabled.

**Testing instructions**

 - Open the console
 - Add a notice

```js
wp.data.dispatch('core/notices').createNotice(
		'info',
		'Post published.',
		{
			type: 'snackbar',
			actions: [
				{
					onClick: () => {},
					label: 'View post'
				}
			]
		}
	);
```

youknowriad committed:
Add media query hooks and animation to snackbar notices

gziolo commented:
![Screen Shot 2019-05-30 at 11 41 08](https://user-images.githubusercontent.com/699132/58624125-d67f0300-82cf-11e9-9a2f-59f53023906f.png)

How much of this size gets bundled in `@wordpress/compose`? Did you measure? After gzip, it might have very little impact, but it would be great to check it.

youknowriad commented:
I didn't add any dependency to `compose` so I don't expect any change in bundle size. react-spring is `9.7KB` and that is going to be added to the components package.

gziolo commented:
> I didn't add any dependency to `compose` so I don't expect any change in bundle size. react-spring is `9.7KB` and that is going to be added to the components package.

Thanks for follow-up, it makes a lot of sense. I really like how `@wordpress/compose` is now structured to contain both hooks and HOCs 💯

youknowriad committed:
Try to force reduced motion in tests environments

kjellr commented:
This is excellent. The modal slides in from the bottom, and does a simple fade out on removal. Here's a (slightly slowed down) GIF of the animation: 

![animation](https://user-images.githubusercontent.com/1202812/58649089-39c16300-82d9-11e9-970f-056fa5cc8b65.gif)

It respects the reduce motion media query too, and just appears/disappears in that case: 

![animation-reduce](https://user-images.githubusercontent.com/1202812/58649115-4776e880-82d9-11e9-87dc-6f880d0ad827.gif)

---

The only issue I'm seeing occurs when there are multiple notices, and the bottommost ones are dismissed first: 

![animation-multiple](https://user-images.githubusercontent.com/1202812/58649277-b3595100-82d9-11e9-8344-de5edf21c061.gif)

The notices near the top slide down as you'd expect, but they do a little pause before they settle into place at the very bottom. Is there anything we can do to eliminate that?

mtias commented on packages/compose/src/hooks/use-reduced-motion/index.js:
```diff
@@ -0,0 +1,16 @@
+ *
+ * @return {boolean} Reduced motion preference value.
+ */
+const useReducedMotion =
```
This should come in handy.

mtias commented on packages/compose/src/hooks/use-media-query/index.js:
```diff
@@ -0,0 +1,24 @@
+ * @param {string} query Media Query.
+ * @return {boolean} return value of the media query.
+ */
+export default function useMediaQuery( query ) {
```
How would we handle these for mobile native conceptually?

youknowriad replied:
That's a good question, I think Mobile shouldn't really care about that hook. Ultimately, this hook might not be used directly by component. instead we'd use higher-level ones that can be "forked" by mobile for example `useReducedMotion` can be adapted for mobile.

mtias commented on packages/components/src/snackbar/list.js:
```diff
@@ -23,17 +30,39 @@ function SnackbarList( { notices, className, children, onRemove = noop } ) {
+				await next( { height: 0 } );
+			},
+			config: { tension: 300 },
+			immediate: isReducedMotion,
```
I wonder if we should wrap around react-spring primitives and offer this by default.

youknowriad replied:
I'm not sure. Because there is not a unique hook for react spring animations so basically this means creating a custom hook for each one of the react-spring hooks.

youknowriad committed:
Fix the leave animation

youknowriad commented:
Pushed some updates here:

 - I introduced a new `FORCE_REDUCED_MOTION` to be able to force reduced motion for e2e test builds. I'd have preferred to avoid it and instead find a way to change the preference using a runtime code but it doesn't seem like there's a way to do it.

 - I fixed the "leave" animation (cc @kjellr), there's still a `1px` jump at the end of the animation but I spent some time on it and I can't understand where it's coming from. I think it's not very noticeable enough to block this PR. 

What do you all think? What's remaining here.

jasmussen commented:
> there's still a 1px jump at the end of the animation but I spent some time on it and I can't understand where it's coming from

I'm unable to look right at this moment, but it sounds like the change from rendering on the GPU (which happens when you use translate) to rendering on the CPU (default state). This is usually mostly visible in Safari. If this is the case, keeping the item on the GPU forever should fix it, and can be done be applying `will-change: transform;` to the element that gets animated.

The only thing to be mindful of is that this can cause some trickiness when overflow is involved, and even then it's rare.

youknowriad commented:
@jasmussen Actually, I think it's if something happens when the DOM element containing the snackbar gets removed from the DOM. Before its removal, it is supposed to have a height of `0` which means removing it shouldn't have a visual impact but you do notice a very small jump.

youknowriad commented:
I'd appreciate reviews and ✅ if you think it's all good.

kjellr commented:
> I fixed the "leave" animation (cc @kjellr), there's still a 1px jump at the end of the animation but I spent some time on it and I can't understand where it's coming from. I think it's not very noticeable enough to block this PR.

@youknowriad — I think I found a quick fix to this one. Setting the `tension` back to the default `170` value (or removing that line entirely) removes this 1px jump. Not sure why, but it seems to work for me, and the animation still feels just fine. 🙂

I am seeing this PR crash GB in Safari — not sure if that's just a problem on my end. [Here's the error](https://cloudup.com/cDKSyeUXVr7) I'm seeing.

nerrad commented on packages/components/src/snackbar/list.js:
```diff
@@ -20,20 +27,43 @@ import Snackbar from './';
+					<div
+						className="components-snackbar-list__notice-container"
+						ref={ ( ref ) => ref && refMap.set( notice, ref ) }
+					>
```
Why is this an additional nested div?  Could the props and ref just be added to the `<animated.div>` component?

youknowriad replied:
No we can't, that was my original ideal but the way this animation work required the animation divs to be adjacent without any margin between them (otherwise the animation is not smooth on removal, there's a jump) and it also requires that the height of the animated div is equal to the snackbar + the margin between two snackbars and without the inner div it's impossible to do because of the "margin-collapsing"

nerrad commented on packages/compose/src/index.js:
```diff
@@ -20,3 +15,15 @@ export { default as withState } from './with-state';
+
+// Hooks
+export { default as useMediaQuery } from './hooks/use-media-query';
+export { default as useReducedMotion } from './hooks/use-reduced-motion';
```
Nice little bit of cleanup here 👍

nerrad commented on packages/components/package.json:
```diff
@@ -44,6 +44,7 @@
 		"re-resizable": "^4.7.1",
 		"react-click-outside": "^3.0.0",
 		"react-dates": "^17.1.1",
+		"react-spring": "^8.0.20",
```
Is `react-spring` something we might consider adding as a library in WP core?  I imagine a number of plugin/theme author's might use it as well (I'm currently using it for a block I'm developing).

youknowriad replied:
I personally worry about adding this kind of libs as WP core libs because of the breaking changes they might introduce. For the moment I think it's small enough to "allow it" to be duplicated but I think that's a good question and we should probably have some defined guidelines to answer it.

nerrad reviewed (approved):
I have just a few non-blocking comments (note this was a code review only, I did not test this).

youknowriad committed:
Fix safari bug and the 1px jump

kjellr commented:
> Good find. Both are fixed now.

Excellent. This should be ready to 🚢 from a design perspective!
